xheap.py

```diff
@@ -114,31 +114,22 @@ def push(self, item):
         _siftdown(self, 0, len(self)-1)
 
     def pop(self, index=None):
+        index = index or 0
+        if index < 0:
+            index += len(self)
+        if index < 0:
+            raise IndexError
         return_item = self._pop(index)
         del self._indexes[return_item]
         return return_item
 
     def _pop(self, index):
-        if index:
-            last_item = super(Heap, self).pop()
-            if index == len(self):
-                return_value = last_item
-            else:
-                return_value = self[index]
-                self[index] = last_item
-                if self[(index - 1) >> 1] < last_item:
```

Comment:
in your code in case of  `self[(index - 1) >> 1] < last_item` evaluates to True the _siftup is executed, in my code the the _siftdown operation exits as soon as it find that condition. 
```
In [13]: from heapq import _siftdown

In [14]: _siftdown??
Signature: _siftdown(heap, startpos, pos)
Source:
def _siftdown(heap, startpos, pos):
    newitem = heap[pos]
    # Follow the path to the root, moving parents down until finding a place
    # newitem fits.
    while pos > startpos:
        parentpos = (pos - 1) >> 1
        parent = heap[parentpos]
        if cmp_lt(newitem, parent):  # this evaluated to false and it breaks out
            heap[pos] = parent
            pos = parentpos
            continue
        break
    heap[pos] = newitem
File:      /usr/lib/python2.7/heapq.py
Type:      function
```

similarly in other case _siftup breaks out in the first iteration. these lines are simply doing redundant checks which are already performed by the _siftup and _siftdown.